geoh5py/shared/concatenation/drillholes_concatenator.py

```diff
@@ -0,0 +1,271 @@
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public License
+#  along with geoh5py.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+from typing import Any
+from uuid import UUID
+
+import numpy as np
+
+from ...groups import DrillholeGroup
+from .concatenator import Concatenator
+from .data import ConcatenatedData
+from .property_group import ConcatenatedPropertyGroup
+
+
+class DrillholesConcatenator(Concatenator, DrillholeGroup):
+    """
+    Class for concatenating drillhole data.
+    """
+
+    def get_data_from_name(self, name: str) -> ConcatenatedData:
+        """
+        Get the data from a given name.
+
+        :param name: The name of the data to extract.
+
+        :return: The first data found with the given name in index.
+        """
+
+        # ensure data_name is in the data
+        if name not in self.data:
+            raise KeyError(f"Data '{name}' not found in concatenated data.")
+
+        # get the data of the know association
+        data: ConcatenatedData = self.workspace.get_entity(  # type: ignore
+            UUID(self.index[name][0][-2].decode("utf-8").strip("{}"))
+        )[0].get_data(UUID(self.index[name][0][-1].decode("utf-8").strip("{}")))[0]
+
+        return data
+
+    @staticmethod
+    def get_depth_association(
+        property_group: ConcatenatedPropertyGroup,
+    ) -> tuple[str] | tuple[str, str] | None:
+        """
+        Based on a PropertyGroup, it gets the name of the depth (or from to)
+            associated with the data.
+
+        :param property_group: The property group to extract.
+
+        return: the name of the values to used for association.
+        """
+        if property_group.property_group_type == "Interval table":
+            return property_group.from_.name, property_group.to_.name
+        if property_group.property_group_type == "Depth table":
+            return (property_group.depth_.name,)
+
+        return None
+
+    def association_by_drillhole(
+        self,
+        names: tuple,
+    ) -> dict:
+        """
+        Based on the first name of the input list,
+        get for every object index and count of all the data in 'names'
+
+        :param names: The names of the data to extract.
+
+        :return: A dictionary with the object uuid and the index of all the data.
+        """
+        # ensure all the names are in the index
+        if not all(name in self.index for name in names):
+            raise KeyError(f"Data '{names}' not found in concatenated data.")
+
+        object_index = {
+            drillhole: {
+                name: list(
+                    self.index[name][self.index[name]["Object ID"] == drillhole][0]
+                )[:2]
+                for name in names
```

Comment:
We can avoid the nested loop if we assume single element.

Reply:
Well, the nested loop is taking account To if from-to to avoid different process for depth and from to

For from to, only from is considerated
For depth, only depth